response/core/serializers.py

```diff
@@ -8,36 +8,61 @@
     class Meta:
         model = ExternalUser
-        fields = ('app_id', 'external_id', 'owner', 'display_name')
+        fields = ("app_id", "external_id", "display_name")
 
 
-class ActionSerializer(serializers.HyperlinkedModelSerializer):
+class ActionSerializer(serializers.ModelSerializer):
     # Serializers define the API representation.
-    incident = serializers.PrimaryKeyRelatedField(queryset=Incident.objects.all(), required=False)
-    user = serializers.PrimaryKeyRelatedField(queryset=ExternalUser.objects.all(), required=False)
+    incident = serializers.PrimaryKeyRelatedField(
+        queryset=Incident.objects.all(), required=False
+    )
+    user = serializers.PrimaryKeyRelatedField(
+        queryset=ExternalUser.objects.all(), required=False
+    )
 
     class Meta:
         model = Action
-        fields = ('pk', 'details', 'done', 'incident', 'user')
+        fields = ("pk", "details", "done", "incident", "user")
 
 
-class IncidentSerializer(serializers.HyperlinkedModelSerializer):
-    reporter = serializers.PrimaryKeyRelatedField(queryset=ExternalUser.objects.all(), required=False)
-    lead = serializers.PrimaryKeyRelatedField(queryset=ExternalUser.objects.all(), required=False)
+class IncidentSerializer(serializers.ModelSerializer):
+    reporter = ExternalUserSerializer(read_only=True)
+    lead = ExternalUserSerializer()
 
     class Meta:
         model = Incident
-        fields = ('pk','report', 'reporter', 'lead', 'start_time', 'end_time', 'report_time', 'action_set')
+        fields = (
+            "action_set",
+            "end_time",
+            "impact",
+            "lead",
+            "pk",
+            "report",
+            "report_time",
+            "reporter",
+            "severity",
+            "start_time",
+            "summary",
+        )
+
+    def update(self, instance, validated_data):
+        instance.end_time = validated_data.get("end_time", instance.end_time)
+        instance.impact = validated_data.get("impact", instance.impact)
+
+        new_lead = validated_data.get("lead", None)
+        if new_lead:
+            instance.lead = ExternalUser.objects.get(
+                display_name=new_lead["display_name"],
+                external_id=new_lead["external_id"],
+            )
 
-    def __init__(self, *args, **kwargs):
-        super(IncidentSerializer, self).__init__(*args, **kwargs)
-        request = kwargs['context']['request']
-        expand = request.GET.get('expand', "").split(',')
+        instance.report = validated_data.get("report", instance.report)
+        instance.start_time = validated_data.get("start_time", instance.start_time)
+        instance.summary = validated_data.get("summary", instance.summary)
+        instance.severity = validated_data.get("severity", instance.severity)
 
-        if 'actions' in expand:
-            self.fields['action_set'] = ActionSerializer(many=True, read_only=True)
+        instance.save()
+        return instance
```

Comment:
Added a new update method here to control exactly how we're deserializing a PUT request and updating the incident instance.

I removed the previous `__init__` function as I don't think we need this functionality, though might well put it back in when I get to the Actions API